### PR TITLE
Fix whitespace column datetime coercion for pandas < 0.17

### PR DESCRIPTION
**Pull request by kwmsmith**
This patches `pandas.to_datetime()` for pandas < 0.17.  Does not affect things for Pandas == 0.17.

Resolves issue #340.

### CODE REVIEW

File: odo/backends/pandas.py [hunk 1]
```diff
@@ -57,9 +57,18 @@ def coerce_datetimes(df):
...
-        partial(pd.to_datetime, errors='ignore')
-    )
+
+    def patched_to_datetime(s):
+        # NOTE: In pandas < 0.17, pd.to_datetime(' ') == datetime(...), which
+        # is not what we want.  This patched version detects empty or
+        # whitespace-only strings, and maps them to a dummy object.  This
+        # prevents datetime coercion on whitespace or empty strings.
+        o = object()
+        s_norm = s.apply(lambda x: o if (hasattr(x, 'strip') and not x.strip()) else x)
```
Comment (jreback) [inline odo/backends/pandas.py]: use s.str.strip() is much faster that using apply

Comment (kwmsmith) [inline odo/backends/pandas.py]: Holy crap, that's awesome.  Thanks for the pointer.

Comment (jreback) [inline odo/backends/pandas.py]: also don't apply pd.to_datetime it works the other way
eg pd.to_datetime(s) is vectorized

File: odo/backends/pandas.py [hunk 2]
```diff
@@ -57,9 +57,20 @@ def coerce_datetimes(df):
     name            object
     dtype: object
     """
-    df2 = df.select_dtypes(include=['object']).apply(
-        partial(pd.to_datetime, errors='ignore')
-    )
+
+    def patched_to_datetime(s):
```
Comment (cpcloud) [inline odo/backends/pandas.py]: I think a cleaner way to do this is something like the following:

```python
objects = df.select_dtypes(include=['object'])
df2 = objects[[c for c in objects.columns if not np.any(objects[c].str.isspace() | objects[c].str.isalpha())]].apply(partial(pd.to_datetime, errors='ignore'))
```

Comment (cpcloud) [inline odo/backends/pandas.py]: This is the same logic that we use for inferring datetime values in `datashape.discover`

File: odo/backends/pandas.py [hunk 3]
```diff
@@ -57,9 +57,20 @@ def coerce_datetimes(df):
...
+        # whitespace-only strings, and maps them to a dummy object.  This
+        # prevents datetime coercion on whitespace or empty strings.
+        o = object()
+        try:
+            s[s.str.strip() == ''] = o
+        except TypeError:
+            pass
+        return pd.to_datetime(s, errors='ignore')
+
+    df2 = df.select_dtypes(include=['object']).apply(patched_to_datetime)
```
Comment (jreback) [inline odo/backends/pandas.py]: you ought to tests this. This i *very* inefficient. ``pd.to_datetime`` is by definition a vectorized operation ITSELF.

Comment (kwmsmith) [inline odo/backends/pandas.py]: For pandas < 0.17, `to_datetime` is not vectorized on dataframes, only on timeseries.  I checked.

So this is the best we can do, vectorization-wise, for Pandas < 0.17, unless you know something I don't.

**Comment by kwmsmith** [comment]
@jreback @cpcloud I think this is ready to go -- thoughts?

**Comment by jreback** [comment]
@kwmsmith ``pd.to_datetime`` is has always been vectorized by definition. It is ONLY applicable to Series.

you are applying to values!

e.g. doing something like ``df[[.....]].apply(pd.to_datetime)`` is ok as you are applying column by column
by ``df[column].apply(pd.to_datetime)`` is a no-no

**Comment by cpcloud** [comment]
@jreback @kwmsmith isn't apply to values we're doing something similar to your `df[[...]].apply` example

**Comment by jreback** [comment]
oh ok then!

File: odo/backends/tests/test_csv.py [hunk 4]
```diff
@@ -65,6 +65,13 @@ def test_pandas_read_supports_datetimes():
         assert df.dtypes['when'] == 'M8[ns]'
 
 
+def test_pandas_read_supports_whitespace_strings():
```
Comment (cpcloud) [inline odo/backends/tests/test_csv.py]: i don't think this test fails on current master

Comment (cpcloud) [inline odo/backends/tests/test_csv.py]: i just checked out the PR and reverted the fix and this test still passes

Comment (kwmsmith) [inline odo/backends/tests/test_csv.py]: What version of Pandas?  If I revert the fix and run the tests with Pandas `0.16.2`, it fails, as it should.

Pandas 0.17 fixes the `pd.to_datetime()` bug that we're working around here.

```
❯❯❯ py.test odo/backends/tests/test_csv_ws.py
========================================================= test session starts =========================================================
platform darwin -- Python 2.7.10, pytest-2.8.1, py-1.4.30, pluggy-0.3.1
rootdir: /Users/ksmith/work/odo, inifile:
collected 41 items

odo/backends/tests/test_csv.py ....F....................................

============================================================== FAILURES ===============================================================
____________________________________________ test_pandas_read_supports_whitespace_strings _____________________________________________

    def test_pandas_read_supports_whitespace_strings():
        with filetext('a,b, \n1,2, \n2,3, \n', extension='csv') as fn:
            csv = CSV(fn)
            ds = discover(csv)
>           assert ds == datashape.dshape("var * {a: int64, b: int64, '': ?string}")
E           assert dshape("var *...: ?datetime}") == dshape("var * ...'': ?string}")
E             Use -v to get the full diff

odo/backends/tests/test_csv.py:72: AssertionError
================================================= 1 failed, 40 passed in 0.36 seconds =================================================
```

**Comment by cpcloud** [comment]
right ... sorry about that

**Comment by cpcloud** [comment]
merge when ready

**Pull request merged**